# Credit Default Risk Prediction
# Part 2. Feature Engineering

**Table of Contents**<a id='toc0_'></a>    
- 1. [Introduction   ](#toc1_)    
- 2. [Notebook preparation   ](#toc2_)    
- 3. [Feature Engineering   ](#toc3_)    
  - 3.1. [Main Home Credit Applications   ](#toc3_1_)    
  - 3.2. [Previous Credits and Their Balance from Other Institutions ](#toc3_2_)    
  - 3.3. [Previous Applications for Home Credit Loans ](#toc3_3_)    
  - 3.4. [Balance of Previous Point of Sales and Cash Loans from Home Credit ](#toc3_4_)    
  - 3.5. [Balance of Previous Credit Cards from Home Credit ](#toc3_5_)    
  - 3.6. [Installments on Previous Loans from Home Credit ](#toc3_6_)    
  - 3.7. [Summary on Feature Engineering](#toc3_7_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=false
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

Note that the links in the above table of contents do not work on GitHub or Google Colab. In Google Colab, please use its embedded table of contents.

## 1. <a id='toc1_'></a>Introduction    [&#8593;](#toc0_)

This notebook is dedicated to feature engineering for the credit default risk prediction model. The focus is on enhancing the predictive capabilities of the model through informed feature engineering. It involves refining the dataset to capture underlying relationships and patterns crucial for predicting credit default risk. To begin, the challenge of disparate currencies within the dataset is addressed by transforming absolute monetary values into ratios. This ensures consistency and comparability across different currency denominations, facilitating more accurate modeling. Drawing upon insights from the experiences of predecessors tackling similar tasks, a feature aimed at identifying the mean target behavior of comparable applicants is created. Additionally, various aggregations are applied across all data tables, with diverse-span windows employed to get meaningful insights and uncover predictive patterns. 

## 2. <a id='toc2_'></a>Notebook preparation    [&#8593;](#toc0_)

Import libraries for data wrangling.

In [1]:
import pandas as pd
import polars as pl
import numpy as np
from polars import col as c
import polars.selectors as cs
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer, make_column_selector
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import TargetEncoder, OneHotEncoder
from feature_engine.selection import DropFeatures
from helper_functions import credit_data_utils
from helper_functions.feature_utils import (
    ApplicationsFeatureCreation,
    NNFeature,
    get_drop_multicollinear,
    get_drop_by_boruta,
)
import joblib
import warnings

warnings.filterwarnings("ignore")

## 3. <a id='toc3_'></a>Feature Engineering    [&#8593;](#toc0_)

All checking on whether the engineered features should be kept is done on training dataset.

In [2]:
idx_train = pl.read_csv("idx_train.csv", has_header=False)

### 3.1. <a id='toc3_1_'></a>Main Home Credit Applications    [&#8593;](#toc0_)

Based on exploratory data analysis only columns related to target and columns selected for feature engineering are loaded from main applications table.

In [3]:
columns = [
    "SK_ID_CURR",
    "TARGET",
    "NAME_CONTRACT_TYPE",
    "CODE_GENDER",
    "AMT_GOODS_PRICE",
    "NAME_INCOME_TYPE",
    "NAME_EDUCATION_TYPE",
    "NAME_FAMILY_STATUS",
    "NAME_HOUSING_TYPE",
    "REGION_POPULATION_RELATIVE",
    "DAYS_BIRTH",
    "DAYS_EMPLOYED",
    "DAYS_REGISTRATION",
    "DAYS_ID_PUBLISH",
    "OWN_CAR_AGE",
    "FLAG_WORK_PHONE",
    "FLAG_PHONE",
    "OCCUPATION_TYPE",
    "REGION_RATING_CLIENT_W_CITY",
    "HOUR_APPR_PROCESS_START",
    "REG_CITY_NOT_LIVE_CITY",
    "REG_CITY_NOT_WORK_CITY",
    "ORGANIZATION_TYPE",
    "EXT_SOURCE_1",
    "EXT_SOURCE_2",
    "EXT_SOURCE_3",
    "APARTMENTS_AVG",
    "YEARS_BUILD_AVG",
    "FLOORSMAX_AVG",
    "COMMONAREA_AVG",
    "DEF_60_CNT_SOCIAL_CIRCLE",
    "DAYS_LAST_PHONE_CHANGE",
    "FLAG_DOCUMENT_3",
    "AMT_INCOME_TOTAL",
    "AMT_ANNUITY",
    "AMT_CREDIT",
    "CNT_CHILDREN",
    "CNT_FAM_MEMBERS",
]
appls = (
    pl.read_csv("./data/application_train.csv")
    .filter(c.SK_ID_CURR.is_in(idx_train))
    .pipe(credit_data_utils.initial_application_cleaning)
    .select(columns)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
)
appls = appls.with_columns(
    CODE_GENDER=pl.when(c.CODE_GENDER == "M")
    .then(pl.lit(0))
    .when(c.CODE_GENDER == "F")
    .then(pl.lit(1))
    .otherwise(pl.lit(None))
    .cast(pl.Float32),
    DAYS_EMPLOYED=c.DAYS_EMPLOYED.fill_null(0),
    OWN_CAR_AGE=c.OWN_CAR_AGE.fill_null(0),
)
for col in appls.columns:
    column = appls[col]
    if ("NAME_" in col) | ("_TYPE" in col) | ("WALLSMATERIAL_" in col):
        appls = appls.with_columns(column.fill_null("XNA").alias(col))
appls_join = {
    "other": appls.select(["SK_ID_CURR", "TARGET"]),
    "on": "SK_ID_CURR",
    "how": "left",
}
appls = appls.to_pandas()
X_appls = appls.drop(columns=["TARGET", "SK_ID_CURR"])
y_appls = appls["TARGET"]
del appls

Size before cleaning: 415.89 MB
Initial number of columns: 122
Size after cleaning: 329.39 MB
Number of columns left: 94
Size before memory reduction: 178.17 MB
Initial data types Counter({Int64: 17, Float64: 12, String: 8, Enum(categories=['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']): 1})
Size after memory reduction: 29.30 MB
Final data types Counter({Float32: 12, Categorical(ordering='physical'): 8, Int8: 7, Int32: 6, Int16: 4, Enum(categories=['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']): 1})


Preprocessing contains imputation of missing values and encoding of categorical features. Target (label) encoding is selected for that. Imputation of missing values in numerical columns is important to prevent missing value propagation to engineered features. However, this all need to be set in a pipeline to prevent data leakage.

In [4]:
appls_preprocess_pipe = ColumnTransformer(
    [
        (
            "int_imp",
            SimpleImputer(strategy="most_frequent"),
            make_column_selector(pattern="FLAG_|CODE_|REG_|REGION_RATING_"),
        ),
        (
            "cont_imp",
            SimpleImputer(strategy="median"),
            make_column_selector(pattern="DAYS_|AMT_|CNT_|_AVG|EXT_|_RELATIV"),
        ),
        (
            "target_enc",
            TargetEncoder(target_type="binary", random_state=42),
            make_column_selector(dtype_include="category"),
        ),
    ],
    remainder="passthrough",
    verbose_feature_names_out=False,
).set_output(transform="pandas")

Custom transformers are created for feature engineering (see `feature_utils.py`in `helper_functions`). It is not possible to engineer features inplace as missing value imputation should be performed in a pipeline, which is only fitted on training data. A special feature is mean of target for 500 closest neighbors defined by the EXT_SOURCE and CREDIT_ANNUITY_RATIO features as described in this [notebook](https://www.kaggle.com/c/home-credit-default-risk/discussion/64821). The final pipeline for processing applications table includes preprocessing and feature engineering.

In [5]:
appls_full_pipe = Pipeline(
    steps=[
        ("preprocessing", appls_preprocess_pipe),
        ("feature_engineering", ApplicationsFeatureCreation()),
        ("nn_feature", NNFeature()),
    ]
).set_output(transform="pandas")

Multicollinear features are dropped using smart correlation selector, which selects features to drop based on their individual performance with logistic regression model.

In [6]:
X_appls_feat = appls_full_pipe.fit_transform(X_appls, y_appls)
drop_corr_appls = get_drop_multicollinear(
    X_appls_feat, y_appls, "drop_corr_appls", refit=False
)
X_appls_feat = X_appls_feat.drop(columns=drop_corr_appls)

Number of quasi-constant and multicollinear aggregated features to drop: 7


Boruta with random forest estimators is used to drop unimportant features.

In [7]:
drop_boruta_appls = get_drop_by_boruta(
    X_appls_feat, y_appls, "drop_boruta_appls", refit=False
)
X_appls_feat = X_appls_feat.drop(columns=drop_boruta_appls)
appls_col_num = X_appls_feat.shape[1]
print(f"Number of selected features: {appls_col_num}")

Number of unimportant aggregated features to drop using Boruta: 9
Number of selected features: 38


Based on the results, the pipeline can be updated.

In [8]:
appls_full_pipe = Pipeline(
    steps=[
        ("preprocessing", appls_preprocess_pipe),
        ("feature_engineering", ApplicationsFeatureCreation()),
        ("nn_feature", NNFeature()),
        ("drop_corr", DropFeatures(drop_corr_appls + drop_boruta_appls)),
    ]
).set_output(transform="pandas")
joblib.dump(appls_full_pipe, "./job_files/appls_full_pipe.joblib")

['./job_files/appls_full_pipe.joblib']

The final number of features from the main table to be used in modeling is 38.

In [9]:
del X_appls, y_appls, appls_preprocess_pipe

### 3.2. <a id='toc3_2_'></a>Previous Credits and Their Balance from Other Institutions  [&#8593;](#toc0_)

Status from Bureau balance table is encoded with order. Also, weighted status feature is created to put more importance on more recent status.

In [10]:
bureau_balance = pl.read_csv("./data/bureau_balance.csv").pipe(
    credit_data_utils.reduce_memory_usage_pl
)
bureau_balance = bureau_balance.with_columns(
    STATUS=pl.when(c.STATUS == "C")
    .then(pl.lit(0))
    .when(c.STATUS == "0")
    .then(pl.lit(1))
    .when(c.STATUS == "X")
    .then(pl.lit(2))
    .when(c.STATUS == "1")
    .then(pl.lit(3))
    .when(c.STATUS == "2")
    .then(pl.lit(4))
    .when(c.STATUS == "3")
    .then(pl.lit(5))
    .when(c.STATUS == "4")
    .then(pl.lit(6))
    .when(c.STATUS == "5")
    .then(pl.lit(7))
)
bureau_balance = bureau_balance.with_columns(
    WEIGHTED_STATUS=c.STATUS / (-c.MONTHS_BALANCE + 1)
)
bureau_balance.head()

Size before memory reduction: 833.13 MB
Initial data types Counter({Int64: 2, String: 1})
Size after memory reduction: 260.35 MB
Final data types Counter({Int32: 1, Int16: 1, Categorical(ordering='physical'): 1})


shape: (5, 4)
┌──────────────┬────────────────┬────────┬─────────────────┐
│ SK_ID_BUREAU ┆ MONTHS_BALANCE ┆ STATUS ┆ WEIGHTED_STATUS │
│ ---          ┆ ---            ┆ ---    ┆ ---             │
│ i32          ┆ i16            ┆ i32    ┆ f64             │
╞══════════════╪════════════════╪════════╪═════════════════╡
│ 5715448      ┆ 0              ┆ 0      ┆ 0.0             │
│ 5715448      ┆ -1             ┆ 0      ┆ 0.0             │
│ 5715448      ┆ -2             ┆ 0      ┆ 0.0             │
│ 5715448      ┆ -3             ┆ 0      ┆ 0.0             │
│ 5715448      ┆ -4             ┆ 0      ┆ 0.0             │
└──────────────┴────────────────┴────────┴─────────────────┘

Various aggregations related to the last, average or median status are created.

In [11]:
bureau_balance_agg = bureau_balance.group_by("SK_ID_BUREAU").agg(
    MONTHS_BALANCE_LAST=c.MONTHS_BALANCE.max(),
    MONTHS_BALANCE_FIRST=c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_DURATION=c.MONTHS_BALANCE.max() - c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_MEDIAN=c.MONTHS_BALANCE.median(),
    STATUS_LAST=c.STATUS.filter(
        c.MONTHS_BALANCE == c.MONTHS_BALANCE.max()
    ).first(),
    STATUS_AVG=c.STATUS.mean(),
    STATUS_MEDIAN=c.STATUS.median(),
    STATUS_MAX=c.STATUS.max(),
    WEIGHTED_STATUS_LAST=c.WEIGHTED_STATUS.filter(
        c.MONTHS_BALANCE == c.MONTHS_BALANCE.max()
    ).first(),
    WEIGHTED_STATUS_AVG=c.WEIGHTED_STATUS.mean(),
    WEIGHTED_STATUS_MAX=c.WEIGHTED_STATUS.max(),
)
bureau_balance_agg.head()

shape: (5, 12)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_BUR ┆ MONTHS_BA ┆ MONTHS_BA ┆ MONTHS_BA ┆ … ┆ STATUS_MA ┆ WEIGHTED_ ┆ WEIGHTED_ ┆ WEIGHTED │
│ EAU       ┆ LANCE_LAS ┆ LANCE_FIR ┆ LANCE_DUR ┆   ┆ X         ┆ STATUS_LA ┆ STATUS_AV ┆ _STATUS_ │
│ ---       ┆ T         ┆ ST        ┆ ATION     ┆   ┆ ---       ┆ ST        ┆ G         ┆ MAX      │
│ i32       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ i32       ┆ ---       ┆ ---       ┆ ---      │
│           ┆ i16       ┆ i16       ┆ i16       ┆   ┆           ┆ f64       ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 6830985   ┆ -5        ┆ -23       ┆ 18        ┆ … ┆ 2         ┆ 0.0       ┆ 0.03584   ┆ 0.086957 │
│ 6831009   ┆ 0         ┆ -48       ┆ 48        ┆ … ┆ 1         ┆ 0.0       ┆ 0.002173  ┆ 0.022222 │
│ 5003854   ┆ 0         ┆ -42       ┆ 42        ┆ … ┆ 1         ┆ 1.0       ┆ 0.101163  ┆ 1.0      │
│ 5205339   ┆ 0         ┆ -4        ┆ 4         ┆ … ┆ 1         ┆ 1.0       ┆ 0.456667  ┆ 1.0      │
│ 5029197   ┆ 0         ┆ -66       ┆ 66        ┆ … ┆ 2         ┆ 0.0       ┆ 0.001612  ┆ 0.029851 │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Aggregated Bureau balance table is joined to Bureau table for further feature engineering. Some ratio columns are created initially.

In [12]:
bureau = (
    pl.read_csv("./data/bureau.csv")
    .filter(c.SK_ID_CURR.is_in(idx_train))
    .pipe(credit_data_utils.initial_bureau_cleaning)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
    .join(other=bureau_balance_agg, on="SK_ID_BUREAU", how="left")
)
bureau = bureau.with_columns(
    DEBT_CREDIT_RATIO=c.AMT_CREDIT_SUM_DEBT / (c.AMT_CREDIT_SUM + 1),
    OVERDUE_CREDIT_RATIO=c.AMT_CREDIT_SUM_OVERDUE / (c.AMT_CREDIT_SUM + 1),
    MAX_OVERDUE_CREDIT_RATIO=c.AMT_CREDIT_MAX_OVERDUE / (c.AMT_CREDIT_SUM + 1),
    LIMIT_CREDIT_RATIO=c.AMT_CREDIT_SUM_LIMIT / (c.AMT_CREDIT_SUM + 1),
)
bureau.head()

Size before memory reduction: 191.91 MB
Initial data types Counter({Int64: 8, Float64: 6, String: 3})
Size after memory reduction: 64.41 MB
Final data types Counter({Float32: 6, Int16: 4, Int32: 3, Categorical(ordering='physical'): 3, Int8: 1})


shape: (5, 32)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_CUR ┆ SK_ID_BUR ┆ CREDIT_AC ┆ CREDIT_CU ┆ … ┆ DEBT_CRED ┆ OVERDUE_C ┆ MAX_OVERD ┆ LIMIT_CR │
│ R         ┆ EAU       ┆ TIVE      ┆ RRENCY    ┆   ┆ IT_RATIO  ┆ REDIT_RAT ┆ UE_CREDIT ┆ EDIT_RAT │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ IO        ┆ _RATIO    ┆ IO       │
│ i32       ┆ i32       ┆ cat       ┆ cat       ┆   ┆ f64       ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆ f64       ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 215354    ┆ 5714462   ┆ Closed    ┆ currency  ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ null     │
│           ┆           ┆           ┆ 1         ┆   ┆           ┆           ┆           ┆          │
│ 215354    ┆ 5714463   ┆ Active    ┆ currency  ┆ … ┆ 0.761517  ┆ 0.0       ┆ 0.0       ┆ null     │
│           ┆           ┆           ┆ 1         ┆   ┆           ┆           ┆           ┆          │
│ 215354    ┆ 5714464   ┆ Active    ┆ currency  ┆ … ┆ null      ┆ 0.0       ┆ 0.0       ┆ null     │
│           ┆           ┆           ┆ 1         ┆   ┆           ┆           ┆           ┆          │
│ 215354    ┆ 5714465   ┆ Active    ┆ currency  ┆ … ┆ null      ┆ 0.0       ┆ 0.0       ┆ null     │
│           ┆           ┆           ┆ 1         ┆   ┆           ┆           ┆           ┆          │
│ 215354    ┆ 5714466   ┆ Active    ┆ currency  ┆ … ┆ null      ┆ 0.0       ┆ 0.028768  ┆ null     │
│           ┆           ┆           ┆ 1         ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

A dedicated function for Bureau data aggregation is created to enable aggregation on groups filtered by specific categorical columns.

In [13]:
def bureau_aggregations_func(
    df: pl.DataFrame, suffix: str = "", left_df: pl.DataFrame = None
) -> pl.DataFrame:
    """Function to aggregate bureau data and optionally add suffix to new
    columns or join with other dataframe."""
    agg_df = df.group_by("SK_ID_CURR").agg(
        cs.contains("DAYS_").mean().round().name.suffix("_AVG"),
        cs.contains("DAYS_").median().name.suffix("_MEDIAN"),
        cs.contains("DAYS_").max().name.suffix("_LAST"),
        cs.contains(["STATUS_", "MONTHS_", "_RATIO"])
        .max()
        .name.suffix("_MAX"),
        cs.contains(["STATUS_", "MONTHS_", "_RATIO"])
        .min()
        .name.suffix("_MIN"),
        cs.contains(["STATUS_", "MONTHS_", "_RATIO"])
        .mean()
        .name.suffix("_AVG"),
        CNT_CREDIT=c.SK_ID_CURR.count(),
        CNT_PROLONG_TOTAL=c.CNT_CREDIT_PROLONG.sum(),
        CREDIT_DAY_OVERDUE_TOTAL=c.CREDIT_DAY_OVERDUE.sum(),
        CREDIT_DAY_OVERDUE_MAX=c.CREDIT_DAY_OVERDUE.max(),
        DEBT_CREDIT_SUM_RATIO=(c.AMT_CREDIT_SUM_DEBT.sum())
        / (c.AMT_CREDIT_SUM.sum() + 1),
        OVERDUE_CREDIT_SUM_RATIO=(c.AMT_CREDIT_SUM_OVERDUE.sum())
        / (c.AMT_CREDIT_SUM.sum() + 1),
        MAX_OVERDUE_CREDIT_SUM_RATIO=(c.AMT_CREDIT_MAX_OVERDUE.sum())
        / (c.AMT_CREDIT_SUM.sum() + 1),
        LIMIT_CREDIT_SUM_RATIO=(c.AMT_CREDIT_SUM_LIMIT.sum())
        / (c.AMT_CREDIT_SUM.sum() + 1),
    )
    agg_df = agg_df.rename(
        lambda col: col + suffix if col != "SK_ID_CURR" else col
    )
    if left_df is not None:
        agg_df = left_df.join(other=agg_df, on="SK_ID_CURR", how="left")
    return agg_df

Finally, the function is applied and all created features are joined based on application ID.

In [14]:
bureau_agg = bureau.group_by("SK_ID_CURR").agg(
    CNT_CREDIT_TYPES_ALL=c.CREDIT_TYPE.n_unique(),
    CNT_CREDIT_TYPES_ACTIVE=c.CREDIT_TYPE.filter(
        c.CREDIT_ACTIVE == "Active"
    ).n_unique(),
    CNT_CREDIT_TYPES_CLOSED=c.CREDIT_TYPE.filter(
        c.CREDIT_ACTIVE == "Closed"
    ).n_unique(),
    ACTIVE_CREDIT_RATIO_ALL=(c.DAYS_CREDIT_ENDDATE > 0).mean(),
    ACTIVE_CREDIT_RATIO_CARD=(c.DAYS_CREDIT_ENDDATE > 0)
    .filter(c.CREDIT_TYPE == "Credit card")
    .mean(),
    ACTIVE_CREDIT_RATIO_CONSUMER=(c.DAYS_CREDIT_ENDDATE > 0)
    .filter(c.CREDIT_TYPE == "Consumer credit")
    .mean(),
)
bureau_agg = bureau_aggregations_func(bureau, "_ALL", bureau_agg)
bureau_agg = bureau.filter(c.CREDIT_ACTIVE == "Active").pipe(
    bureau_aggregations_func, "_ACTIVE", bureau_agg
)
bureau_agg = bureau.filter(c.CREDIT_ACTIVE == "Closed").pipe(
    bureau_aggregations_func, "_CLOSED", bureau_agg
)
bureau_agg = bureau.filter(c.CREDIT_TYPE == "Credit card").pipe(
    bureau_aggregations_func, "_CARD", bureau_agg
)
bureau_agg = bureau.filter(c.CREDIT_TYPE == "Consumer credit").pipe(
    bureau_aggregations_func, "_CONSUMER", bureau_agg
)
bureau_agg = bureau.filter(
    (c.CREDIT_ACTIVE == "Active") & (c.CREDIT_TYPE == "Credit card")
).pipe(bureau_aggregations_func, "_ACTIVE_CARD", bureau_agg)
bureau_agg = bureau.filter(
    (c.CREDIT_ACTIVE == "Active") & (c.CREDIT_TYPE == "Consumer credit")
).pipe(bureau_aggregations_func, "_ACTIVE_CONSUMER", bureau_agg)
bureau_agg = bureau.filter(
    (c.CREDIT_ACTIVE == "Closed") & (c.CREDIT_TYPE == "Credit card")
).pipe(bureau_aggregations_func, "_CLOSED_CARD", bureau_agg)
bureau_agg = bureau.filter(
    (c.CREDIT_ACTIVE == "Closed") & (c.CREDIT_TYPE == "Consumer credit")
).pipe(bureau_aggregations_func, "_CLOSED_CONSUMER", bureau_agg)

To enable initial removal of multicollinear and unimportant features, missing values are simply filled with 0.

In [15]:
bureau_agg_filled = (
    bureau_agg.filter(c.SK_ID_CURR.is_in(idx_train))
    .fill_null(value=0)
    .fill_nan(value=0)
    .join(**appls_join)
    .to_pandas()
)
X_bureau = bureau_agg_filled.drop(columns=["TARGET", "SK_ID_CURR"])
y_bureau = bureau_agg_filled["TARGET"]

In [16]:
drop_corr_bureau = get_drop_multicollinear(
    X_bureau, y_bureau, "drop_corr_bureau", refit=False
)
X_bureau = X_bureau.drop(columns=drop_corr_bureau)

Number of quasi-constant and multicollinear aggregated features to drop: 349


In [17]:
drop_boruta_bureau = get_drop_by_boruta(
    X_bureau, y_bureau, "drop_boruta_bureau", refit=False
)
X_bureau = X_bureau.drop(columns=drop_boruta_bureau)
bureau_col_num = X_bureau.shape[1]
print(f"Number of selected features: {bureau_col_num}")

Number of unimportant aggregated features to drop using Boruta: 133
Number of selected features: 109


Finally, 109 aggregated features for Bureau and its balance data are saved for later use. Note, that missing values are not imputed with 0 in the saved table.

In [18]:
bureau_agg.drop(drop_corr_bureau + drop_boruta_bureau).write_csv(
    "./data_agg/bureau_agg.csv"
)

In [19]:
del (
    bureau_balance,
    bureau_balance_agg,
    bureau,
    bureau_agg,
    bureau_agg_filled,
    X_bureau,
    y_bureau,
)

### 3.3. <a id='toc3_3_'></a>Previous Applications for Home Credit Loans  [&#8593;](#toc0_)

Only non-collinear columns as found during EDA is loaded and those useful for feature engineering. Only the last applications per individual contract are used for feature engineering. Missing values in categorical columns are encoded with XNA. Features describing differences and ratios of amounts are created first. A potentially very important feature related to interest of the loan is also engineered. However, it tends to be negative for revolving loans. In that case, the interest feature is set to 0.

In [20]:
columns_drop = [
    "RATE_INTEREST_PRIMARY",
    "RATE_INTEREST_PRIVILEGED",
    "DAYS_FIRST_DRAWING",
    "NAME_TYPE_SUITE",
    "WEEKDAY_APPR_PROCESS_START",
    "NAME_GOODS_CATEGORY",
    "SELLERPLACE_AREA",
    "DAYS_LAST_DUE_1ST_VERSION",
    "NFLAG_LAST_APPL_IN_DAY",
    "FLAG_LAST_APPL_PER_CONTRACT",
]
prev = (
    pl.read_csv("./data/previous_application.csv")
    .pipe(credit_data_utils.initial_prev_application_cleaning)
    .filter(c.FLAG_LAST_APPL_PER_CONTRACT == 1)
    .drop(columns_drop)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
)
for col in prev.columns:
    column = prev[col]
    if (
        ("NAME_" in col)
        | ("_TYPE" in col)
        | ("CODE_" in col)
        | ("_COMBINATION" in col)
    ):
        prev = prev.with_columns(column.fill_null("XNA").alias(col))
prev = prev.with_columns(
    INTEREST=((c.CNT_PAYMENT * c.AMT_ANNUITY) - c.AMT_CREDIT).clip(
        lower_bound=0
    )  # credit card interest ends up being 0
)
prev = prev.with_columns(
    INTEREST_RATE=c.INTEREST / (c.AMT_CREDIT * c.CNT_PAYMENT + 1),
    INTEREST_SHARE=(c.INTEREST / (c.AMT_CREDIT + 1)).clip(upper_bound=1),
    CREDIT_ANNUITY_RATIO=c.AMT_CREDIT / (c.AMT_ANNUITY + 1),
    ANNUITY_CREDIT_RATIO=c.AMT_ANNUITY / (c.AMT_CREDIT + 1),
    GOODS_CREDIT_RATIO=c.AMT_GOODS_PRICE / (c.AMT_CREDIT + 1),
    GOODS_CREDIT_DIFF=c.AMT_GOODS_PRICE - c.AMT_CREDIT,
    GOODS_APPLICATION_RATIO=c.AMT_GOODS_PRICE / (c.AMT_APPLICATION + 1),
    GOODS_APPLICATION_DIFF=c.AMT_GOODS_PRICE - c.AMT_APPLICATION,
    APPLICATION_CREDIT_RATIO=c.AMT_APPLICATION / (c.AMT_CREDIT + 1),
    APPLICATION_CREDIT_DIFF=c.AMT_APPLICATION - c.AMT_CREDIT,
)
prev.head()

Size before cleaning: 776.58 MB
Size after cleaning: 759.95 MB
Number of columns left: 37
Size before memory reduction: 578.87 MB
Initial data types Counter({String: 12, Int64: 8, Float64: 6, Enum(categories=['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']): 1})
Size after memory reduction: 154.35 MB
Final data types Counter({Categorical(ordering='physical'): 12, Float32: 6, Int16: 5, Int32: 2, Enum(categories=['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']): 1, Int8: 1})


shape: (5, 38)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ NAME_CONT ┆ AMT_ANNUI ┆ … ┆ GOODS_APP ┆ GOODS_APP ┆ APPLICATI ┆ APPLICAT │
│ V         ┆ R         ┆ RACT_TYPE ┆ TY        ┆   ┆ LICATION_ ┆ LICATION_ ┆ ON_CREDIT ┆ ION_CRED │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ RATIO     ┆ DIFF      ┆ _RATIO    ┆ IT_DIFF  │
│ i32       ┆ i32       ┆ cat       ┆ f32       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆ f32       ┆ f64       ┆ f32      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2030495   ┆ 271877    ┆ Consumer  ┆ 1730.4300 ┆ … ┆ 0.999942  ┆ 0.0       ┆ 0.999942  ┆ 0.0      │
│           ┆           ┆ loans     ┆ 54        ┆   ┆           ┆           ┆           ┆          │
│ 2802425   ┆ 108129    ┆ Cash      ┆ 25188.615 ┆ … ┆ 0.999998  ┆ 0.0       ┆ 0.893813  ┆ -72171.0 │
│           ┆           ┆ loans     ┆ 234       ┆   ┆           ┆           ┆           ┆          │
│ 2523466   ┆ 122040    ┆ Cash      ┆ 15060.735 ┆ … ┆ 0.999991  ┆ 0.0       ┆ 0.824505  ┆ -23944.5 │
│           ┆           ┆ loans     ┆ 352       ┆   ┆           ┆           ┆           ┆          │
│ 2819243   ┆ 176158    ┆ Cash      ┆ 47041.335 ┆ … ┆ 0.999998  ┆ 0.0       ┆ 0.955838  ┆ -20790.0 │
│           ┆           ┆ loans     ┆ 938       ┆   ┆           ┆           ┆           ┆          │
│ 1784265   ┆ 202054    ┆ Cash      ┆ 31924.394 ┆ … ┆ 0.999997  ┆ 0.0       ┆ 0.83528   ┆ -66555.0 │
│           ┆           ┆ loans     ┆ 531       ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

To enable aggregations of categorical features they are one-hot-encoded.

In [21]:
def custom_combiner(feature, category):
    if str(category) == "infrequent_sklearn":
        name = "OHE_" + str(feature).split("_", 1)[-1] + "_RARE"
    else:
        name = (
            "OHE_"
            + str(feature).split("_", 1)[-1]
            + "_"
            + str(category)
            .upper()
            .replace("/ ", "")
            .replace(" ", "_")
            .replace("-", "_")
        )
    return name


cat_columns = [
    col
    for col in prev.columns
    if any(sub in col for sub in ["NAME_", "PRODUCT_", "CODE_", "_TYPE"])
]
ohe_pipe = ColumnTransformer(
    [
        (
            "ohe",
            OneHotEncoder(
                min_frequency=0.05,
                sparse_output=False,
                dtype=np.int8,
                feature_name_combiner=custom_combiner,
            ),
            cat_columns,
        ),
    ],
    remainder="passthrough",
    verbose_feature_names_out=False,
).set_output(transform="polars")
ohe_pipe.fit(prev.filter(c.SK_ID_CURR.is_in(idx_train)))
prev = prev.select(["SK_ID_PREV"] + cat_columns).join(
    other=ohe_pipe.transform(prev),
    on="SK_ID_PREV",
    how="left",
)
prev.head()

shape: (5, 90)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ NAME_CONT ┆ NAME_CASH ┆ NAME_CONT ┆ … ┆ GOODS_APP ┆ GOODS_APP ┆ APPLICATI ┆ APPLICAT │
│ V         ┆ RACT_TYPE ┆ _LOAN_PUR ┆ RACT_STAT ┆   ┆ LICATION_ ┆ LICATION_ ┆ ON_CREDIT ┆ ION_CRED │
│ ---       ┆ ---       ┆ POSE      ┆ US        ┆   ┆ RATIO     ┆ DIFF      ┆ _RATIO    ┆ IT_DIFF  │
│ i32       ┆ cat       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆ cat       ┆ cat       ┆   ┆ f64       ┆ f32       ┆ f64       ┆ f32      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2030495   ┆ Consumer  ┆ XAP       ┆ Approved  ┆ … ┆ 0.999942  ┆ 0.0       ┆ 0.999942  ┆ 0.0      │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2802425   ┆ Cash      ┆ XNA       ┆ Approved  ┆ … ┆ 0.999998  ┆ 0.0       ┆ 0.893813  ┆ -72171.0 │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2523466   ┆ Cash      ┆ XNA       ┆ Approved  ┆ … ┆ 0.999991  ┆ 0.0       ┆ 0.824505  ┆ -23944.5 │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2819243   ┆ Cash      ┆ XNA       ┆ Approved  ┆ … ┆ 0.999998  ┆ 0.0       ┆ 0.955838  ┆ -20790.0 │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 1784265   ┆ Cash      ┆ Repairs   ┆ Refused   ┆ … ┆ 0.999997  ┆ 0.0       ┆ 0.83528   ┆ -66555.0 │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Again, function is written to enable aggregations over data grouped by different columns.

In [22]:
def previous_aggregations_func(
    df: pl.DataFrame, suffix: str = "", left_df: pl.DataFrame = None
) -> pl.DataFrame:
    """Function to aggregate previous applications data and optionally add
    suffix to new columns or join with other dataframe."""
    agg_df = df.group_by("SK_ID_CURR").agg(
        cs.contains("OHE_").sum().name.suffix("_CNT"),
        cs.contains(["DAYS_", "CNT_"]).mean().round().name.suffix("_AVG"),
        cs.contains(["DAYS_", "CNT_"]).median().name.suffix("_MEDIAN"),
        cs.contains(["DAYS_", "CNT_"]).max().name.suffix("_MAX"),
        cs.contains(["AMT_", "RATE_", "_RATIO", "_DIFF", "INTEREST_"])
        .max()
        .name.suffix("_MAX"),
        cs.contains(["AMT_", "RATE_", "_RATIO", "_DIFF", "INTEREST_"])
        .min()
        .name.suffix("_MIN"),
        cs.contains(["AMT_", "RATE_", "_RATIO", "_DIFF", "INTEREST_"])
        .mean()
        .name.suffix("_AVG"),
        CNT_APPLS=c.SK_ID_CURR.count(),
    )
    agg_df = agg_df.rename(
        lambda col: col + suffix if col != "SK_ID_CURR" else col
    )
    if left_df is not None:
        agg_df = left_df.join(other=agg_df, on="SK_ID_CURR", how="left")
    return agg_df

Finally aggregations over previous application data are run and joined to form a dataset of 964 features.

In [23]:
prev_agg = prev.group_by("SK_ID_CURR").agg(
    (cs.all() - cs.contains(["SK_ID_", "OHE_"]))
    .filter(c.DAYS_DECISION == c.DAYS_DECISION.max())
    .first()
    .name.suffix("_LAST")
)
prev_agg = previous_aggregations_func(prev, "_ALL", prev_agg)
prev_agg = prev.filter(c.NAME_CONTRACT_TYPE == "Consumer loans").pipe(
    previous_aggregations_func, "_CONSUMER", prev_agg
)
prev_agg = prev.filter(c.NAME_CONTRACT_TYPE == "Cash loans").pipe(
    previous_aggregations_func, "_CASH", prev_agg
)
prev_agg = prev.filter(c.NAME_CONTRACT_TYPE == "Revolving loans").pipe(
    previous_aggregations_func, "_REVOLVING", prev_agg
)
prev_agg = prev.filter(c.NAME_CONTRACT_STATUS == "Approved").pipe(
    previous_aggregations_func, "_APPROVED", prev_agg
)
prev_agg = prev.filter(c.NAME_CONTRACT_STATUS == "Refused").pipe(
    previous_aggregations_func, "_REFUSED", prev_agg
)
prev_agg = (
    prev.sort(c.DAYS_DECISION, descending=True)
    .group_by("SK_ID_CURR")
    .head(5)
    .pipe(previous_aggregations_func, "_5LAST", prev_agg)
)
prev_agg = (
    prev.sort(c.DAYS_DECISION, descending=True)
    .group_by("SK_ID_CURR")
    .head(3)
    .pipe(previous_aggregations_func, "_3LAST", prev_agg)
)
prev_agg.head()

shape: (5, 965)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_CUR ┆ NAME_CONT ┆ NAME_CASH ┆ NAME_CONT ┆ … ┆ GOODS_APP ┆ APPLICATI ┆ APPLICATI ┆ CNT_APPL │
│ R         ┆ RACT_TYPE ┆ _LOAN_PUR ┆ RACT_STAT ┆   ┆ LICATION_ ┆ ON_CREDIT ┆ ON_CREDIT ┆ S_3LAST  │
│ ---       ┆ _LAST     ┆ POSE_LAST ┆ US_LAST   ┆   ┆ DIFF_AVG_ ┆ _RATIO_AV ┆ _DIFF_AVG ┆ ---      │
│ i32       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ 3LAST     ┆ G_3LA…    ┆ _3LAS…    ┆ u32      │
│           ┆ cat       ┆ cat       ┆ cat       ┆   ┆ ---       ┆ ---       ┆ ---       ┆          │
│           ┆           ┆           ┆           ┆   ┆ f32       ┆ f64       ┆ f32       ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 332209    ┆ Cash      ┆ XNA       ┆ Refused   ┆ … ┆ 0.0       ┆ 0.880319  ┆ -56356.5  ┆ 3        │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 289991    ┆ Revolving ┆ XAP       ┆ Approved  ┆ … ┆ 0.0       ┆ 0.96553   ┆ -18571.5  ┆ 3        │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 111610    ┆ Revolving ┆ XAP       ┆ Refused   ┆ … ┆ 0.0       ┆ 0.999957  ┆ 0.0       ┆ 3        │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 273235    ┆ Cash      ┆ XNA       ┆ Canceled  ┆ … ┆ 0.0       ┆ 0.278241  ┆ -21681.0  ┆ 3        │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 411122    ┆ Cash      ┆ XNA       ┆ Refused   ┆ … ┆ 0.0       ┆ 0.89609   ┆ -90501.0  ┆ 3        │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Since there are still categorical features in the dataset, target encoding is used to enable elimination of multicollinear and unimportant features.

In [24]:
enc_pipe = ColumnTransformer(
    [
        (
            "target_enc",
            TargetEncoder(target_type="binary", random_state=42),
            make_column_selector(dtype_include="category"),
        ),
    ],
    remainder="passthrough",
    verbose_feature_names_out=False,
).set_output(transform="pandas")

prev_agg_filled = (
    prev_agg.filter(c.SK_ID_CURR.is_in(idx_train))
    .fill_null(value=0)
    .fill_nan(value=0)
    .join(**appls_join)
    .to_pandas()
)
X_prev = prev_agg_filled.drop(columns=["TARGET", "SK_ID_CURR"])
y_prev = prev_agg_filled["TARGET"]
X_prev = enc_pipe.fit_transform(X_prev, y_prev)

In [25]:
drop_corr_prev = get_drop_multicollinear(
    X_prev, y_prev, "drop_corr_prev", refit=False
)
X_prev = X_prev.drop(columns=drop_corr_prev)

Number of quasi-constant and multicollinear aggregated features to drop: 519


In [26]:
drop_boruta_prev = get_drop_by_boruta(
    X_prev, y_prev, "drop_boruta_prev", refit=False
)
X_prev = X_prev.drop(columns=drop_boruta_prev)
prev_col_num = X_prev.shape[1]
print(f"Number of selected features: {prev_col_num}")

Number of unimportant aggregated features to drop using Boruta: 174
Number of selected features: 271


Finally, 271 aggregated feature for previous application data is saved for later use. Note, that missing values are not imputed with 0 and categorical features are not encoded in the saved table.

In [27]:
prev_agg.drop(drop_corr_prev + drop_boruta_prev).write_csv(
    "./data_agg/prev_agg.csv"
)

In [28]:
del (
    prev,
    prev_agg,
    prev_agg_filled,
    X_prev,
    y_prev,
)

### 3.4. <a id='toc3_4_'></a>Balance of Previous Point of Sales and Cash Loans from Home Credit  [&#8593;](#toc0_)

Before aggregation, missing values in a single categorical column from POS and cash loan balance data are filled with XNA.

In [29]:
pos_cash = (
    pl.read_csv("./data/POS_CASH_balance.csv")
    .pipe(credit_data_utils.initial_pos_cash_credit_card_installments_cleaning)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
)
pos_cash = pos_cash.with_columns(c.NAME_CONTRACT_STATUS.fill_null("XNA"))
pos_cash.head()

Size before cleaning: 689.24 MB
Size after cleaning: 690.43 MB
Number of columns left: 8
Size before memory reduction: 690.43 MB
Initial data types Counter({Int64: 7, String: 1})
Size after memory reduction: 251.57 MB
Final data types Counter({Int32: 4, Int16: 3, Categorical(ordering='physical'): 1})


shape: (5, 8)
┌────────────┬────────────┬────────────┬────────────┬────────────┬────────────┬────────┬───────────┐
│ SK_ID_PREV ┆ SK_ID_CURR ┆ MONTHS_BAL ┆ CNT_INSTAL ┆ CNT_INSTAL ┆ NAME_CONTR ┆ SK_DPD ┆ SK_DPD_DE │
│ ---        ┆ ---        ┆ ANCE       ┆ MENT       ┆ MENT_FUTUR ┆ ACT_STATUS ┆ ---    ┆ F         │
│ i32        ┆ i32        ┆ ---        ┆ ---        ┆ E          ┆ ---        ┆ i32    ┆ ---       │
│            ┆            ┆ i16        ┆ i16        ┆ ---        ┆ cat        ┆        ┆ i32       │
│            ┆            ┆            ┆            ┆ i16        ┆            ┆        ┆           │
╞════════════╪════════════╪════════════╪════════════╪════════════╪════════════╪════════╪═══════════╡
│ 1803195    ┆ 182943     ┆ -31        ┆ 48         ┆ 45         ┆ Active     ┆ 0      ┆ 0         │
│ 1715348    ┆ 367990     ┆ -33        ┆ 36         ┆ 35         ┆ Active     ┆ 0      ┆ 0         │
│ 1784872    ┆ 397406     ┆ -32        ┆ 12         ┆ 9          ┆ Active     ┆ 0      ┆ 0         │
│ 1903291    ┆ 269225     ┆ -35        ┆ 48         ┆ 42         ┆ Active     ┆ 0      ┆ 0         │
│ 2341044    ┆ 334279     ┆ -35        ┆ 36         ┆ 35         ┆ Active     ┆ 0      ┆ 0         │
└────────────┴────────────┴────────────┴────────────┴────────────┴────────────┴────────┴───────────┘

First some aggregations are done on individual loans to have information on mean, median, sum, last, minimum or maximum values.

In [30]:
pos_cash_agg_initial = pos_cash.group_by(["SK_ID_PREV", "SK_ID_CURR"]).agg(
    cs.contains(["SK_DPD", "NAME_", "MONTHS_", "_FUTURE"])
    .filter(c.MONTHS_BALANCE == c.MONTHS_BALANCE.max())
    .first()
    .name.suffix("_LAST"),
    cs.contains("SK_DPD").max().name.suffix("_MAX"),
    cs.contains("SK_DPD").sum().name.suffix("_SUM"),
    cs.contains("SK_DPD").mean().round().name.suffix("_AVG"),
    MONTHS_BALANCE_FIRST=c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_DURATION=c.MONTHS_BALANCE.max() - c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_MEDIAN=c.MONTHS_BALANCE.median(),
    CNT_INSTALMENT_MEDIAN=c.CNT_INSTALMENT.median(),
)
pos_cash_agg_initial.head()

shape: (5, 17)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ MONTHS_BA ┆ CNT_INSTA ┆ … ┆ MONTHS_BA ┆ MONTHS_BA ┆ MONTHS_BA ┆ CNT_INST │
│ V         ┆ R         ┆ LANCE_LAS ┆ LMENT_FUT ┆   ┆ LANCE_FIR ┆ LANCE_DUR ┆ LANCE_MED ┆ ALMENT_M │
│ ---       ┆ ---       ┆ T         ┆ URE_LAST  ┆   ┆ ST        ┆ ATION     ┆ IAN       ┆ EDIAN    │
│ i32       ┆ i32       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆ i16       ┆ i16       ┆   ┆ i16       ┆ i16       ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 1715348   ┆ 367990    ┆ -16       ┆ 0         ┆ … ┆ -34       ┆ 18        ┆ -25.0     ┆ 36.0     │
│ 2341044   ┆ 334279    ┆ -1        ┆ 1         ┆ … ┆ -36       ┆ 35        ┆ -18.5     ┆ 36.0     │
│ 2207092   ┆ 342166    ┆ -20       ┆ 0         ┆ … ┆ -32       ┆ 12        ┆ -26.0     ┆ 12.0     │
│ 1387235   ┆ 153211    ┆ -1        ┆ 2         ┆ … ┆ -35       ┆ 34        ┆ -18.0     ┆ 36.0     │
│ 1716688   ┆ 405549    ┆ -32       ┆ 0         ┆ … ┆ -44       ┆ 12        ┆ -38.0     ┆ 12.0     │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Again, function is written to apply aggregations over different groups of data.

In [31]:
def pos_cash_aggregations_func(
    df: pl.DataFrame, suffix: str = "", left_df: pl.DataFrame = None
) -> pl.DataFrame:
    """Function to aggregate POS and cash balance data and optionally add
    suffix to new columns or join with other dataframe."""
    agg_df = df.group_by("SK_ID_CURR").agg(
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .mean()
        .round()
        .name.suffix("_AVG"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .median()
        .name.suffix("_MEDIAN"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .max()
        .name.suffix("_MAX"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .min()
        .name.suffix("_MIN"),
        CNT=c.SK_ID_CURR.count(),
    )
    agg_df = agg_df.rename(
        lambda col: col + suffix if col != "SK_ID_CURR" else col
    )
    if left_df is not None:
        agg_df = left_df.join(other=agg_df, on="SK_ID_CURR", how="left")
    return agg_df

These aggregations are applied over the whole dataset and also only on the information from the last two years.

In [32]:
pos_cash_agg = pos_cash_aggregations_func(pos_cash_agg_initial, "_ALL")
pos_cash_agg = pos_cash_agg_initial.filter(
    c.NAME_CONTRACT_STATUS_LAST == "Completed"
).pipe(pos_cash_aggregations_func, "_COMPLETE", pos_cash_agg)
pos_cash_agg = pos_cash_agg_initial.filter(
    c.NAME_CONTRACT_STATUS_LAST != "Completed"
).pipe(pos_cash_aggregations_func, "_NONCOMPLETE", pos_cash_agg)
pos_cash_agg = pos_cash_agg_initial.filter(c.MONTHS_BALANCE_LAST >= -24).pipe(
    pos_cash_aggregations_func, "_ALL_2YRS", pos_cash_agg
)
pos_cash_agg = pos_cash_agg_initial.filter(
    (c.MONTHS_BALANCE_LAST >= -24)
    & (c.NAME_CONTRACT_STATUS_LAST == "Completed")
).pipe(pos_cash_aggregations_func, "_COMPLETE_2YRS", pos_cash_agg)
pos_cash_agg = pos_cash_agg_initial.filter(
    (c.MONTHS_BALANCE_LAST >= -24)
    & (c.NAME_CONTRACT_STATUS_LAST != "Completed")
).pipe(pos_cash_aggregations_func, "_NONCOMPLETE_2YRS", pos_cash_agg)
pos_cash_agg = pos_cash_agg.with_columns(
    COMPLETE_ALL_RATIO=c.CNT_COMPLETE / c.CNT_ALL,
    COMPLETE_ALL_2YRS_RATIO=c.CNT_COMPLETE_2YRS / c.CNT_ALL_2YRS,
    NONCOMPLETE_ALL_RATIO=c.CNT_NONCOMPLETE / c.CNT_ALL,
    NONCOMPLETE_ALL_2YRS_RATIO=c.CNT_NONCOMPLETE_2YRS / c.CNT_ALL_2YRS,
)
pos_cash_agg.head()

shape: (5, 347)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_CUR ┆ MONTHS_BA ┆ CNT_INSTA ┆ SK_DPD_LA ┆ … ┆ COMPLETE_ ┆ COMPLETE_ ┆ NONCOMPLE ┆ NONCOMPL │
│ R         ┆ LANCE_LAS ┆ LMENT_FUT ┆ ST_AVG_AL ┆   ┆ ALL_RATIO ┆ ALL_2YRS_ ┆ TE_ALL_RA ┆ ETE_ALL_ │
│ ---       ┆ T_AVG_ALL ┆ URE_LAST_ ┆ L         ┆   ┆ ---       ┆ RATIO     ┆ TIO       ┆ 2YRS_RAT │
│ i32       ┆ ---       ┆ AVG_A…    ┆ ---       ┆   ┆ f64       ┆ ---       ┆ ---       ┆ IO       │
│           ┆ f64       ┆ ---       ┆ f64       ┆   ┆           ┆ f64       ┆ f64       ┆ ---      │
│           ┆           ┆ f64       ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 376243    ┆ -9.0      ┆ 2.0       ┆ 0.0       ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.5      │
│ 394473    ┆ -11.0     ┆ 2.0       ┆ 0.0       ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.5      │
│ 356411    ┆ -6.0      ┆ 0.0       ┆ 0.0       ┆ … ┆ 1.0       ┆ 1.0       ┆ null      ┆ null     │
│ 255255    ┆ -46.0     ┆ 3.0       ┆ 0.0       ┆ … ┆ 0.5       ┆ null      ┆ 0.5       ┆ 1.0      │
│ 182341    ┆ -31.0     ┆ 7.0       ┆ 0.0       ┆ … ┆ 0.5       ┆ 0.666667  ┆ 0.5       ┆ 0.333333 │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

To enable initial removal of multicollinear and unimportant features, missing values are simply filled with 0.

In [33]:
pos_cash_agg_filled = (
    pos_cash_agg.filter(c.SK_ID_CURR.is_in(idx_train))
    .fill_null(value=0)
    .fill_nan(value=0)
    .join(**appls_join)
    .to_pandas()
)
X_pos_cash = pos_cash_agg_filled.drop(columns=["TARGET", "SK_ID_CURR"])
y_pos_cash = pos_cash_agg_filled["TARGET"]

In [34]:
drop_corr_pos_cash = get_drop_multicollinear(
    X_pos_cash, y_pos_cash, "drop_corr_pos_cash", refit=False
)
X_pos_cash = X_pos_cash.drop(columns=drop_corr_pos_cash)

Number of quasi-constant and multicollinear aggregated features to drop: 266


In [35]:
drop_boruta_pos_cash = get_drop_by_boruta(
    X_pos_cash, y_pos_cash, "drop_boruta_pos_cash", refit=False
)
X_pos_cash = X_pos_cash.drop(columns=drop_boruta_pos_cash)
pos_cash_col_num = X_pos_cash.shape[1]
print(f"Number of selected features: {pos_cash_col_num}")

Number of unimportant aggregated features to drop using Boruta: 3
Number of selected features: 77


Finally, 77 aggregated features for POS and cash loan balance data are saved for later use. Note, that missing values are not imputed with 0 in the saved table.

In [36]:
pos_cash_agg.drop(drop_corr_pos_cash + drop_boruta_pos_cash).write_csv(
    "./data_agg/pos_cash_agg.csv"
)

In [37]:
del (
    pos_cash,
    pos_cash_agg,
    pos_cash_agg_initial,
    pos_cash_agg_filled,
    X_pos_cash,
    y_pos_cash,
)

### 3.5. <a id='toc3_5_'></a>Balance of Previous Credit Cards from Home Credit  [&#8593;](#toc0_)

First, various ratio features are created to diminish the impact of absolute values. To enable this, missing values are simply imputed with 0.

In [38]:
card = (
    pl.read_csv("./data/credit_card_balance.csv")
    .pipe(credit_data_utils.initial_pos_cash_credit_card_installments_cleaning)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
)
card = card.fill_null(0).with_columns(
    BALANCE_LIMIT_RATIO=c.AMT_BALANCE / (c.AMT_CREDIT_LIMIT_ACTUAL + 1),
    DRAWINGS_ATM_RATIO=c.AMT_DRAWINGS_ATM_CURRENT
    / (c.AMT_DRAWINGS_CURRENT + 1),
    DRAWINGS_OTHER_RATIO=c.AMT_DRAWINGS_OTHER_CURRENT
    / (c.AMT_DRAWINGS_CURRENT + 1),
    DRAWINGS_POS_RATIO=c.AMT_DRAWINGS_POS_CURRENT
    / (c.AMT_DRAWINGS_CURRENT + 1),
    PAYMENT_INST_RATIO=c.AMT_PAYMENT_CURRENT / (c.AMT_INST_MIN_REGULARITY + 1),
    PAYMENT_TOTAL_INST_RATIO=c.AMT_PAYMENT_TOTAL_CURRENT
    / (c.AMT_INST_MIN_REGULARITY + 1),
    PRINCIPAL_TOTAL_RECEIVABLE_RATIO=c.AMT_RECEIVABLE_PRINCIPAL
    / (c.AMT_TOTAL_RECEIVABLE + 1),
)
card.head()

Size before cleaning: 707.31 MB
Size after cleaning: 707.31 MB
Number of columns left: 23
Size before memory reduction: 707.31 MB
Initial data types Counter({Int64: 11, Float64: 11, String: 1})
Size after memory reduction: 278.80 MB
Final data types Counter({Float32: 11, Int16: 7, Int32: 3, Int8: 1, Categorical(ordering='physical'): 1})


shape: (5, 30)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ MONTHS_BA ┆ AMT_BALAN ┆ … ┆ DRAWINGS_ ┆ PAYMENT_I ┆ PAYMENT_T ┆ PRINCIPA │
│ V         ┆ R         ┆ LANCE     ┆ CE        ┆   ┆ POS_RATIO ┆ NST_RATIO ┆ OTAL_INST ┆ L_TOTAL_ │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ _RATIO    ┆ RECEIVAB │
│ i32       ┆ i32       ┆ i16       ┆ f32       ┆   ┆ f64       ┆ f64       ┆ ---       ┆ LE_RATIO │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆ f64       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2562384   ┆ 378907    ┆ -6        ┆ 56.970001 ┆ … ┆ 0.998862  ┆ 1.057999  ┆ 1.057999  ┆ 0.0      │
│ 2582071   ┆ 363914    ┆ -1        ┆ 63975.554 ┆ … ┆ 0.0       ┆ 0.999556  ┆ 0.999556  ┆ 0.927532 │
│           ┆           ┆           ┆ 688       ┆   ┆           ┆           ┆           ┆          │
│ 1740877   ┆ 371185    ┆ -7        ┆ 31815.224 ┆ … ┆ 0.0       ┆ 0.999556  ┆ 0.999556  ┆ 0.855864 │
│           ┆           ┆           ┆ 609       ┆   ┆           ┆           ┆           ┆          │
│ 1389973   ┆ 337855    ┆ -4        ┆ 236572.10 ┆ … ┆ 0.0       ┆ 1.010871  ┆ 1.010871  ┆ 0.965241 │
│           ┆           ┆           ┆ 9375      ┆   ┆           ┆           ┆           ┆          │
│ 1891521   ┆ 126868    ┆ -1        ┆ 453919.46 ┆ … ┆ 0.999913  ┆ 1.177708  ┆ 1.177708  ┆ 0.97604  │
│           ┆           ┆           ┆ 875       ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Initially, some aggregations are done on individual loans to have information on mean, median, sum, last, minimum or maximum values.

In [39]:
card_agg_initial = card.group_by(["SK_ID_PREV", "SK_ID_CURR"]).agg(
    cs.exclude(["SK_ID_PREV", "SK_ID_CURR"])
    .filter(c.MONTHS_BALANCE == c.MONTHS_BALANCE.max())
    .first()
    .name.suffix("_LAST"),
    cs.contains(["SK_DPD", "CNT_", "AMT_"]).max().name.suffix("_MAX"),
    cs.contains(["SK_DPD", "CNT_", "AMT_"]).sum().name.suffix("_SUM"),
    cs.contains(["SK_DPD", "CNT_", "AMT_"]).mean().round().name.suffix("_AVG"),
    cs.contains(["SK_DPD", "CNT_", "AMT_"]).median().name.suffix("_MEDIAN"),
    cs.contains("_RATIO").max().name.suffix("_MAX"),
    cs.contains("_RATIO").min().name.suffix("_MIN"),
    cs.contains("_RATIO").mean().name.suffix("_AVG"),
    cs.contains("_RATIO").median().name.suffix("_MEDIAN"),
    MONTHS_BALANCE_FIRST=c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_DURATION=c.MONTHS_BALANCE.max() - c.MONTHS_BALANCE.min(),
    MONTHS_BALANCE_MEDIAN=c.MONTHS_BALANCE.median(),
)
card_agg_initial

shape: (104_307, 137)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ MONTHS_BA ┆ AMT_BALAN ┆ … ┆ PRINCIPAL ┆ MONTHS_BA ┆ MONTHS_BA ┆ MONTHS_B │
│ V         ┆ R         ┆ LANCE_LAS ┆ CE_LAST   ┆   ┆ _TOTAL_RE ┆ LANCE_FIR ┆ LANCE_DUR ┆ ALANCE_M │
│ ---       ┆ ---       ┆ T         ┆ ---       ┆   ┆ CEIVABLE_ ┆ ST        ┆ ATION     ┆ EDIAN    │
│ i32       ┆ i32       ┆ ---       ┆ f32       ┆   ┆ RATIO…    ┆ ---       ┆ ---       ┆ ---      │
│           ┆           ┆ i16       ┆           ┆   ┆ ---       ┆ i16       ┆ i16       ┆ f64      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆           ┆           ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2037377   ┆ 289902    ┆ -1        ┆ 131294.34 ┆ … ┆ 0.95133   ┆ -12       ┆ 11        ┆ -6.5     │
│           ┆           ┆           ┆ 375       ┆   ┆           ┆           ┆           ┆          │
│ 1189095   ┆ 231608    ┆ -1        ┆ 0.0       ┆ … ┆ 0.974022  ┆ -32       ┆ 31        ┆ -16.5    │
│ 2392920   ┆ 132481    ┆ -1        ┆ 85792.093 ┆ … ┆ 0.999681  ┆ -7        ┆ 6         ┆ -4.0     │
│           ┆           ┆           ┆ 75        ┆   ┆           ┆           ┆           ┆          │
│ 1927949   ┆ 323730    ┆ -1        ┆ 148956.34 ┆ … ┆ 0.95978   ┆ -76       ┆ 75        ┆ -38.5    │
│           ┆           ┆           ┆ 375       ┆   ┆           ┆           ┆           ┆          │
│ 2257392   ┆ 384410    ┆ -1        ┆ 82818.359 ┆ … ┆ 0.984442  ┆ -11       ┆ 10        ┆ -6.0     │
│           ┆           ┆           ┆ 375       ┆   ┆           ┆           ┆           ┆          │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │
│ 1132920   ┆ 208501    ┆ -1        ┆ 46570.5   ┆ … ┆ 0.999978  ┆ -1        ┆ 0         ┆ -1.0     │
│ 1798731   ┆ 400297    ┆ -2        ┆ 0.0       ┆ … ┆ 0.0       ┆ -2        ┆ 0         ┆ -2.0     │
│ 1887580   ┆ 213062    ┆ -2        ┆ 0.0       ┆ … ┆ 0.0       ┆ -2        ┆ 0         ┆ -2.0     │
│ 1773670   ┆ 424890    ┆ -2        ┆ 228528.67 ┆ … ┆ 0.999996  ┆ -2        ┆ 0         ┆ -2.0     │
│           ┆           ┆           ┆ 1875      ┆   ┆           ┆           ┆           ┆          │
│ 1551072   ┆ 403596    ┆ -1        ┆ 0.0       ┆ … ┆ 0.0       ┆ -1        ┆ 0         ┆ -1.0     │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Again, function is written to apply aggregations over different groups of data.

In [40]:
def card_aggregations_func(
    df: pl.DataFrame, suffix: str = "", left_df: pl.DataFrame = None
) -> pl.DataFrame:
    """Function to aggregate credit card balance data and optionally add
    suffix to new columns or join with other dataframe."""
    agg_df = df.group_by("SK_ID_CURR").agg(
        cs.contains(["MONTHS_", "CNT_", "SK_DPD_"])
        .mean()
        .round()
        .name.suffix("_AVG"),
        cs.contains(["AMT_", "_RATIO"]).mean().name.suffix("_AVG"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .median()
        .name.suffix("_MEDIAN"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .max()
        .name.suffix("_MAX"),
        (cs.all() - cs.contains(["SK_ID_", "NAME_"]))
        .min()
        .name.suffix("_MIN"),
        CNT=c.SK_ID_CURR.count(),
    )
    agg_df = agg_df.rename(
        lambda col: col + suffix if col != "SK_ID_CURR" else col
    )
    if left_df is not None:
        agg_df = left_df.join(other=agg_df, on="SK_ID_CURR", how="left")
    return agg_df

These aggregations are applied over the whole dataset and also only on the information from the last two years, last two months.

In [41]:
card_agg = card_aggregations_func(card_agg_initial, "_ALL")
card_agg = card_agg_initial.filter(
    c.NAME_CONTRACT_STATUS_LAST == "Active"
).pipe(card_aggregations_func, "ACTIVE", card_agg)
card_agg = card.filter(c.MONTHS_BALANCE >= -24).pipe(
    card_aggregations_func, "_2YRS", card_agg
)
card_agg = card.filter(c.MONTHS_BALANCE >= -2).pipe(
    card_aggregations_func, "_2MTHS", card_agg
)

To enable initial removal of multicollinear and unimportant features, missing values are simply filled with 0.

In [42]:
card_agg_filled = (
    card_agg.filter(c.SK_ID_CURR.is_in(idx_train))
    .fill_null(value=0)
    .fill_nan(value=0)
    .join(**appls_join)
    .to_pandas()
)
X_card = card_agg_filled.drop(columns=["TARGET", "SK_ID_CURR"])
y_card = card_agg_filled["TARGET"]

In [43]:
drop_corr_card = get_drop_multicollinear(
    X_card, y_card, "drop_corr_card", refit=False
)
X_card = X_card.drop(columns=drop_corr_card)

Number of quasi-constant and multicollinear aggregated features to drop: 1159


In [44]:
drop_boruta_card = get_drop_by_boruta(
    X_card, y_card, "drop_boruta_card", refit=False
)
X_card = X_card.drop(columns=drop_boruta_card)
card_col_num = X_card.shape[1]
print(f"Number of selected features: {card_col_num}")

Number of unimportant aggregated features to drop using Boruta: 50
Number of selected features: 81


Finally, 81 aggregated feature for credit card balance data is saved for later use. Note, that missing values are not imputed with 0 in the saved table.

In [45]:
card_agg.drop(drop_corr_card + drop_boruta_card).write_csv(
    "./data_agg/card_agg.csv"
)

In [46]:
del (
    card,
    card_agg,
    card_agg_initial,
    card_agg_filled,
    X_card,
    y_card,
)

### 3.6. <a id='toc3_6_'></a>Installments on Previous Loans from Home Credit  [&#8593;](#toc0_)

A few missing values are imputed with 0 to enable feature engineering. First some relative features are created.

In [47]:
inst = (
    pl.read_csv("./data/installments_payments.csv")
    .pipe(credit_data_utils.initial_pos_cash_credit_card_installments_cleaning)
    .pipe(credit_data_utils.reduce_memory_usage_pl)
)
inst = inst.fill_null(0).with_columns(
    DAYS_LATE=c.DAYS_INSTALMENT - c.DAYS_ENTRY_PAYMENT,
    PAYMENT_INSTALMENT_DIFF=c.AMT_PAYMENT - c.AMT_INSTALMENT,
    PAYMENT_INSTALMENT_RATIO=c.AMT_PAYMENT / (c.AMT_INSTALMENT + 1),
)
inst.head()

Size before cleaning: 832.03 MB
Size after cleaning: 832.03 MB
Number of columns left: 8
Size before memory reduction: 832.03 MB
Initial data types Counter({Int64: 6, Float64: 2})
Size after memory reduction: 338.98 MB
Final data types Counter({Int32: 3, Int16: 3, Float32: 2})


shape: (5, 11)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ NUM_INSTA ┆ NUM_INSTA ┆ … ┆ AMT_PAYME ┆ DAYS_LATE ┆ PAYMENT_I ┆ PAYMENT_ │
│ V         ┆ R         ┆ LMENT_VER ┆ LMENT_NUM ┆   ┆ NT        ┆ ---       ┆ NSTALMENT ┆ INSTALME │
│ ---       ┆ ---       ┆ SION      ┆ BER       ┆   ┆ ---       ┆ i32       ┆ _DIFF     ┆ NT_RATIO │
│ i32       ┆ i32       ┆ ---       ┆ ---       ┆   ┆ f32       ┆           ┆ ---       ┆ ---      │
│           ┆           ┆ i16       ┆ i16       ┆   ┆           ┆           ┆ f32       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 1054186   ┆ 161674    ┆ 1         ┆ 6         ┆ … ┆ 6948.3598 ┆ 7         ┆ 0.0       ┆ 0.999856 │
│           ┆           ┆           ┆           ┆   ┆ 63        ┆           ┆           ┆          │
│ 1330831   ┆ 151639    ┆ 0         ┆ 34        ┆ … ┆ 1716.5250 ┆ 0         ┆ 0.0       ┆ 0.999418 │
│           ┆           ┆           ┆           ┆   ┆ 24        ┆           ┆           ┆          │
│ 2085231   ┆ 193053    ┆ 2         ┆ 1         ┆ … ┆ 25425.0   ┆ 0         ┆ 0.0       ┆ 0.999961 │
│ 2452527   ┆ 199697    ┆ 1         ┆ 3         ┆ … ┆ 24350.130 ┆ 8         ┆ 0.0       ┆ 0.999959 │
│           ┆           ┆           ┆           ┆   ┆ 859       ┆           ┆           ┆          │
│ 2714724   ┆ 167756    ┆ 1         ┆ 2         ┆ … ┆ 2160.5849 ┆ -17       ┆ -4.455078 ┆ 0.997482 │
│           ┆           ┆           ┆           ┆   ┆ 61        ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Initially, some aggregations are done on installments for individual loans to have information on mean, median, sum, last, minimum or maximum values.

In [48]:
inst_agg_initial = inst.group_by(["SK_ID_PREV", "SK_ID_CURR"]).agg(
    cs.exclude(["SK_ID_PREV", "SK_ID_CURR"])
    .filter(c.DAYS_INSTALMENT == c.DAYS_INSTALMENT.max())
    .first()
    .name.suffix("_LAST"),
    cs.contains(["DAYS_"])
    .filter(c.DAYS_INSTALMENT == c.DAYS_INSTALMENT.min())
    .first()
    .name.suffix("_FIRST"),
    cs.contains(["AMT_", "DAYS_LATE", "_DIFF", "_RATIO"])
    .max()
    .name.suffix("_MAX"),
    cs.contains(["AMT_", "DAYS_LATE", "_DIFF", "_RATIO"])
    .min()
    .name.suffix("_MIN"),
    cs.contains(["AMT_", "DAYS_LATE", "_DIFF", "_RATIO"])
    .median()
    .name.suffix("_MEDIAN"),
    cs.contains(["AMT_", "_DIFF", "_RATIO"]).mean().name.suffix("_AVG"),
    cs.contains(["AMT_", "DAYS_LATE", "_DIFF"]).sum().name.suffix("_SUM"),
    DAYS_LATE_AVG=c.DAYS_LATE.mean().round(),
    DAYS_INSTALMENT_DURATION=c.DAYS_INSTALMENT.max() - c.DAYS_INSTALMENT.min(),
    DAYS_ENTRY_PAYMENT_DURATION=c.DAYS_ENTRY_PAYMENT.max()
    - c.DAYS_ENTRY_PAYMENT.min(),
    DAYS_INSTALMENT_MEDIAN=c.DAYS_INSTALMENT.median(),
    DAYS_ENTRY_PAYMENT_MEDIAN=c.DAYS_ENTRY_PAYMENT.median(),
)
inst_agg_initial.head()

shape: (5, 42)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_PRE ┆ SK_ID_CUR ┆ NUM_INSTA ┆ NUM_INSTA ┆ … ┆ DAYS_INST ┆ DAYS_ENTR ┆ DAYS_INST ┆ DAYS_ENT │
│ V         ┆ R         ┆ LMENT_VER ┆ LMENT_NUM ┆   ┆ ALMENT_DU ┆ Y_PAYMENT ┆ ALMENT_ME ┆ RY_PAYME │
│ ---       ┆ ---       ┆ SION_LAST ┆ BER_LAST  ┆   ┆ RATION    ┆ _DURATION ┆ DIAN      ┆ NT_MEDIA │
│ i32       ┆ i32       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ N        │
│           ┆           ┆ i16       ┆ i16       ┆   ┆ i16       ┆ i32       ┆ f64       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 1137312   ┆ 164489    ┆ 2         ┆ 15        ┆ … ┆ 420       ┆ 429       ┆ -1504.0   ┆ -1534.0  │
│ 1053286   ┆ 170192    ┆ 2         ┆ 23        ┆ … ┆ 660       ┆ 679       ┆ -335.0    ┆ -335.5   │
│ 2702769   ┆ 102122    ┆ 2         ┆ 8         ┆ … ┆ 210       ┆ 214       ┆ -176.0    ┆ -191.0   │
│ 1840786   ┆ 175114    ┆ 1         ┆ 12        ┆ … ┆ 330       ┆ 325       ┆ -611.0    ┆ -609.0   │
│ 2671380   ┆ 111601    ┆ 1         ┆ 10        ┆ … ┆ 270       ┆ 279       ┆ -585.0    ┆ -596.0   │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

Again, function is written to apply aggregations over different groups of data.

In [49]:
def inst_aggregations_func(
    df: pl.DataFrame, suffix: str = "", left_df: pl.DataFrame = None
) -> pl.DataFrame:
    """Function to aggregate installments payments data and optionally add
    suffix to new columns or join with other dataframe."""
    agg_df = df.group_by("SK_ID_CURR").agg(
        cs.contains(["DAYS_"]).mean().round().name.suffix("_AVG"),
        cs.contains(["AMT_", "_DIFF", "_RATIO"]).mean().name.suffix("_AVG"),
        cs.contains(["AMT_", "_DIFF", "_RATIO", "DAYS_"])
        .median()
        .name.suffix("_MEDIAN"),
        cs.contains(["AMT_", "_DIFF", "_RATIO", "DAYS_"])
        .max()
        .name.suffix("_MAX"),
        cs.contains(["AMT_", "_DIFF", "_RATIO", "DAYS_"])
        .min()
        .name.suffix("_MIN"),
        cs.contains(["AMT_", "_DIFF", "DAYS_LATE"]).sum().name.suffix("_SUM"),
        CNT=c.SK_ID_CURR.count(),
    )
    agg_df = agg_df.rename(
        lambda col: col + suffix if col != "SK_ID_CURR" else col
    )
    if left_df is not None:
        agg_df = left_df.join(other=agg_df, on="SK_ID_CURR", how="left")
    return agg_df

These aggregations are applied over the whole dataset, only credit card installments, only non-credit card installments, only on the installments from the last one and two years.

In [50]:
inst_agg = inst_aggregations_func(inst_agg_initial, "_ALL")
inst_agg = inst_agg_initial.filter(c.NUM_INSTALMENT_VERSION_LAST == 0).pipe(
    inst_aggregations_func, "_CARD", inst_agg
)
inst_agg = inst_agg_initial.filter(c.NUM_INSTALMENT_VERSION_LAST != 0).pipe(
    inst_aggregations_func, "_NONCARD", inst_agg
)
inst_agg = inst.filter(c.DAYS_ENTRY_PAYMENT >= -365).pipe(
    inst_aggregations_func, "_1YR", inst_agg
)
inst_agg = inst.filter(c.DAYS_ENTRY_PAYMENT >= (-365 * 2)).pipe(
    inst_aggregations_func, "_2YRS", inst_agg
)

To enable initial removal of multicollinear and unimportant features, missing values are simply filled with 0.

In [51]:
inst_agg_filled = (
    inst_agg.filter(c.SK_ID_CURR.is_in(idx_train))
    .fill_null(value=0)
    .fill_nan(value=0)
    .join(**appls_join)
    .to_pandas()
)
X_inst = inst_agg_filled.drop(columns=["TARGET", "SK_ID_CURR"])
y_inst = inst_agg_filled["TARGET"]

In [52]:
drop_corr_inst = get_drop_multicollinear(
    X_inst, y_inst, "drop_corr_inst", refit=False
)
X_inst = X_inst.drop(columns=drop_corr_inst)

Number of quasi-constant and multicollinear aggregated features to drop: 410


In [53]:
drop_boruta_inst = get_drop_by_boruta(
    X_inst, y_inst, "drop_boruta_inst", refit=False
)
X_inst = X_inst.drop(columns=drop_boruta_inst)
inst_col_num = X_inst.shape[1]
print(f"Number of selected features: {inst_col_num}")

Number of unimportant aggregated features to drop using Boruta: 26
Number of selected features: 164


Finally, 164 aggregated features for installment data are saved for later use. Note, that missing values are not imputed with 0 in the saved table.

In [54]:
inst_agg.drop(drop_corr_inst + drop_boruta_inst).write_csv(
    "./data_agg/inst_agg.csv"
)

In [55]:
del (
    inst,
    inst_agg,
    inst_agg_initial,
    inst_agg_filled,
    X_inst,
    y_inst,
)

### 3.7. <a id='toc3_7_'></a>Summary on Feature Engineering [&#8593;](#toc0_)

1. A total of 38 features (original and engineered) from the main home credit application table will be further used in modeling. Among them, the mean of the target value of the 500 closest neighbors based on external sources and the credit-annuity ratio will be included.
2. A total of 109 engineered features from the tables on previous credits and their balances from other institutions will be further used in modeling. Aggregations here were applied not only to the whole table but also separately to active and closed credits, as well as to card and consumer credits.
3. A total of 271 engineered features describing the previous applications for Home Credit loans will be further used in modeling. Aggregations here were applied not only to the whole table but also separately for applications of different statuses and types.
4. A total of 77 engineered features describing the balance of previous point-of-sales and cash loans from Home Credit will be further used in modeling. Aggregations were applied to the whole table, to active loans only, and to the last two years.
5. A total of 81 engineered features describing the balance of previous credit cards from Home Credit will be further used in modeling. Aggregations were applied to all credit cards, only to active credit cards, and only to the last two years and the last two months.
6. A total of 164 engineered features describing installments on previous loans from Home Credit will be further used in modeling. Aggregations were applied to the whole table, only to credit card installments, non-credit card installments, and to the installments from the last one year and the last two years.